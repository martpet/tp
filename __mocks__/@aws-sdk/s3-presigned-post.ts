export const createPresignedPost = vi.fn().mockResolvedValue('dummyPresignedPost');
