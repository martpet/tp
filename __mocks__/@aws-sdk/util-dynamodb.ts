export const unmarshall = vi
  .fn()
  .mockName('unmarshall')
  .mockReturnValue('dummyUnmarshalledItems');
