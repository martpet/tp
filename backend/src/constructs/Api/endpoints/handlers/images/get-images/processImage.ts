import { SdkStreamMixin } from '@aws-sdk/types';
import { DetectModerationLabelsCommand, RekognitionClient } from 'lambda-layer';
import sharp from 'sharp';

const rekognition = new RekognitionClient({});

type Props = {
  stream: SdkStreamMixin;
  photoBucket: string;
  s3ObjectName: string;
  quality?: number;
  size?: number;
};

export async function processImage({
  stream,
  photoBucket,
  s3ObjectName,
  size = 1500,
  quality = 100,
}: Props) {
  const moderationCommand = new DetectModerationLabelsCommand({
    Image: { S3Object: { Bucket: photoBucket, Name: s3ObjectName } },
    MinConfidence: 90,
  });
  const { ModerationLabels } = await rekognition.send(moderationCommand);
  const uint8 = await stream.transformToByteArray();
  const image = sharp(uint8).jpeg({ quality }).resize(size);
  const isModerated = ModerationLabels?.some(({ Name }) => Name === 'Explicit Nudity');

  if (isModerated) {
    image.blur(75);
  }

  const buffer = await image.toBuffer();
  const base64 = buffer.toString('base64');
  return base64;
}
