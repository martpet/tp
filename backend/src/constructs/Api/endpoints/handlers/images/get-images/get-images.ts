import {
  APIGatewayProxyHandlerV2,
  apiOptions,
  DynamoDBClient,
  DynamoDBDocumentClient,
  EnvVars,
  errorResponse,
  GetCommand,
  GetImagesResponse,
  GetObjectCommand,
  lambdaPayloadLimit,
  PathParam,
  PhotosTableItem,
  photosTableOptions,
  QueryStrings,
  S3Client,
  StatusCodes,
} from 'lambda-layer';

import { processImage } from './processImage';

const s3Client = new S3Client({});
const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const handler: APIGatewayProxyHandlerV2<GetImagesResponse> = async ({
  queryStringParameters,
  pathParameters,
}) => {
  const { photoBucket } = process.env as EnvVars<'/images', 'GET'>;
  const { fingerprint } = pathParameters as PathParam<'/photos', 'GET'>;
  const { quality, size } = Object(queryStringParameters) as QueryStrings<'/images'>;

  if (!photoBucket) {
    return errorResponse('68e8d0232f');
  }

  if (!fingerprint) {
    return errorResponse('564208f603', { statusCode: StatusCodes.BAD_REQUEST });
  }

  const photoItem = await getPhotoItem(fingerprint);

  if (!photoItem) {
    return errorResponse('5b2797f93a', { statusCode: StatusCodes.NOT_FOUND });
  }

  let stream;
  const s3ObjectName = `${photoItem.userId}/${fingerprint}`;

  try {
    stream = await getS3Stream(photoBucket, s3ObjectName);
  } catch (error) {
    return errorResponse('602d977044', { error });
  }

  if (!stream) {
    return errorResponse('dc1c2c42bf');
  }

  let base64Image;

  try {
    base64Image = await processImage({
      stream,
      photoBucket,
      s3ObjectName,
      size: Number(size) || undefined,
      quality: Number(quality) || undefined,
    });
  } catch (error) {
    return errorResponse('aa6d018607', { error });
  }

  if (base64Image.length > lambdaPayloadLimit) {
    return errorResponse('5fe36afa17');
  }

  return {
    statusCode: StatusCodes.OK,
    isBase64Encoded: true,
    body: base64Image,
    headers: {
      'Content-Type': 'image/jpeg',
      'Cache-Control': `max-age=${apiOptions['/images'].cacheProps.maxTtl}`,
    },
  };
};

async function getPhotoItem(fingerprint: string) {
  const getCommand = new GetCommand({
    TableName: photosTableOptions.tableName,
    Key: { [photosTableOptions.partitionKey.name]: fingerprint },
  });
  const { Item } = await ddbDocClient.send(getCommand);
  const photoItem = Item as PhotosTableItem | undefined;
  return photoItem;
}

async function getS3Stream(photoBucket: string, s3ObjectName: string) {
  const getObjectCommand = new GetObjectCommand({
    Bucket: photoBucket,
    Key: s3ObjectName,
  });
  const { Body } = await s3Client.send(getObjectCommand);
  return Body;
}
