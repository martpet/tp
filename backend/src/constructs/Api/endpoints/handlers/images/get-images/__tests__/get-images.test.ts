import { GetObjectCommand, GetObjectCommandOutput, S3Client } from '@aws-sdk/client-s3';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { lambdaPayloadLimit } from 'lambda-layer';

import {
  itCalls,
  itHasEnvVars,
  itHasPathParam,
  itResolves,
  itResolvesWithError,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';

import { handler } from '../get-images';
import { processImage } from '../processImage';

vi.mock('../processImage');
vi.mock('~/constructs/Api/utils/errorResponse');

const ddbMock = mockClient(DynamoDBDocumentClient);
const s3Mock = mockClient(S3Client);

process.env.photoBucket = 'dummyPhotoBucketName';

const args = [
  {
    pathParameters: {
      fingerprint: 'dummyFingerprint',
    },
    queryStringParameters: {
      quality: '99',
      size: '888',
    },
  },
] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  ddbMock.reset();
  s3Mock.reset();

  ddbMock.on(GetCommand).resolves({
    Item: { userId: 'dummyUserId' },
  });

  s3Mock.on(GetObjectCommand).resolves({
    Body: 'dummyBody',
  } as unknown as GetObjectCommandOutput);
});

describe('get-images', () => {
  itHasEnvVars(['photoBucket'], handler, args);
  itHasPathParam('fingerprint', handler, args);
  itSendsAwsCommand(GetCommand, ddbMock, handler, args);
  itSendsAwsCommand(GetObjectCommand, s3Mock, handler, args);
  itCalls(processImage, handler, args);
  itResolves(handler, args);

  describe('when "size" and "item" query strings do not cast to number', () => {
    const argsClone = structuredClone(args);
    const queryStrings = { ...args[0].queryStringParameters };
    delete queryStrings.size;
    delete queryStrings.quality;
    argsClone[0].queryStringParameters = queryStrings;
    itCalls(processImage, handler, argsClone);
  });

  describe('when "Item" is missing from "GetCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(GetCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });

  describe('when "S3Client" rejects "GetObjectCommand"', () => {
    beforeEach(() => {
      s3Mock.on(GetObjectCommand).rejects('Dummy GetObjectCommand error');
    });
    itResolvesWithError(handler, args);
  });

  describe('when "Body" is missing from "GetObjectCommand" output', () => {
    beforeEach(() => {
      s3Mock.on(GetObjectCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });

  describe('when "processImage" rejects', () => {
    beforeEach(() => {
      vi.mocked(processImage).mockRejectedValueOnce('Dummy processImage error');
    });
    itResolvesWithError(handler, args);
  });

  describe('when "processImage" output length is bigger than lambda payload limit', () => {
    beforeEach(() => {
      vi.mocked(processImage).mockResolvedValue({
        length: lambdaPayloadLimit + 1,
      } as any);
    });
    itResolvesWithError(handler, args);
  });
});
