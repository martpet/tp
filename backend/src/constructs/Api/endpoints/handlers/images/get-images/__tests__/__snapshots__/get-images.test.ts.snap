// Vitest Snapshot v1

exports[`get-images > calls "processImage" with correct args 1`] = `
[
  [
    {
      "photoBucket": "dummyPhotoBucketName",
      "quality": 99,
      "s3ObjectName": "dummyUserId/dummyFingerprint",
      "size": 888,
      "stream": "dummyBody",
    },
  ],
]
`;

exports[`get-images > resolves with a correct value 1`] = `
{
  "body": "dummyBase64EncodedImage",
  "headers": {
    "Cache-Control": "max-age=3153600000",
    "Content-Type": "image/jpeg",
  },
  "isBase64Encoded": true,
  "statusCode": 200,
}
`;

exports[`get-images > sends "GetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Key": {
    "fingerprint": "dummyFingerprint",
  },
  "TableName": "TP-photos",
}
`;

exports[`get-images > sends "GetObjectCommand" from "S3Client" with correct args 1`] = `
{
  "Bucket": "dummyPhotoBucketName",
  "Key": "dummyUserId/dummyFingerprint",
}
`;

exports[`get-images > when "Body" is missing from "GetObjectCommand" output > calls "errorResponse" with correct args 1`] = `
[
  [
    "dc1c2c42bf",
  ],
]
`;

exports[`get-images > when "Body" is missing from "GetObjectCommand" output > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "Item" is missing from "GetCommand" output > calls "errorResponse" with correct args 1`] = `
[
  [
    "5b2797f93a",
    {
      "statusCode": 404,
    },
  ],
]
`;

exports[`get-images > when "Item" is missing from "GetCommand" output > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "S3Client" rejects "GetObjectCommand" > calls "errorResponse" with correct args 1`] = `
[
  [
    "602d977044",
    {
      "error": [Error: Dummy GetObjectCommand error],
    },
  ],
]
`;

exports[`get-images > when "S3Client" rejects "GetObjectCommand" > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "fingerprint" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "564208f603",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`get-images > when "fingerprint" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "photoBucket" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "68e8d0232f",
  ],
]
`;

exports[`get-images > when "photoBucket" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "processImage" output length is bigger than lambda payload limit > calls "errorResponse" with correct args 1`] = `
[
  [
    "5fe36afa17",
  ],
]
`;

exports[`get-images > when "processImage" output length is bigger than lambda payload limit > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "processImage" rejects > calls "errorResponse" with correct args 1`] = `
[
  [
    "aa6d018607",
    {
      "error": "Dummy processImage error",
    },
  ],
]
`;

exports[`get-images > when "processImage" rejects > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images > when "size" and "item" query strings do not cast to number > calls "processImage" with correct args 1`] = `
[
  [
    {
      "photoBucket": "dummyPhotoBucketName",
      "quality": undefined,
      "s3ObjectName": "dummyUserId/dummyFingerprint",
      "size": undefined,
      "stream": "dummyBody",
    },
  ],
]
`;
